File: src/python/command_injection.py
#!/usr/bin/env python3
"""
Command Injection vulnerabilities for Python.
Used for testing SAST tools like Bandit.
"""

import os
import subprocess
import sys

def vulnerable_os_system(user_input):
    """Vulnerable command execution using os.system()."""
    # VULNERABLE: Direct command concatenation
    command = f"ping -c 1 {user_input}"
    os.system(command)

def vulnerable_subprocess_call(user_input):
    """Command injection with subprocess.call()."""
    # VULNERABLE: shell=True with user input
    subprocess.call(f"nslookup {user_input}", shell=True)

def vulnerable_subprocess_popen(filename):
    """Command injection with subprocess.Popen."""
    # VULNERABLE: User input in command arguments
    subprocess.Popen(["cat", filename])

def vulnerable_eval(user_input):
    """Code injection using eval()."""
    # VULNERABLE: Evaluating user input
    result = eval(user_input)
    return result

def vulnerable_exec(user_input):
    """Code injection using exec()."""
    # VULNERABLE: Executing user input
    exec(user_input)

def vulnerable_pickle(pickle_data):
    """Insecure deserialization with pickle."""
    import pickle
    # VULNERABLE: Loading untrusted pickle data
    return pickle.loads(pickle_data)

def vulnerable_yaml(yaml_data):
    """Insecure YAML loading."""
    import yaml
    # VULNERABLE: yaml.load instead of yaml.safe_load
    return yaml.load(yaml_data, Loader=yaml.Loader)

def vulnerable_template(user_input):
    """Template injection."""
    from string import Template
    # VULNERABLE: User input in template
    template = Template(f"Hello $name, you entered: {user_input}")
    return template.substitute(name="User")

def safe_command_execution(ip_address):
    """Safe command execution with input validation."""
    import re

    # SAFE: Validate input before use
    if re.match(r'^[0-9.]+$', ip_address):
        subprocess.call(["ping", "-c", "1", ip_address])

def safe_subprocess(user_input):
    """Safe subprocess without shell."""
    # SAFE: Using list argument without shell=True
    subprocess.run(["echo", user_input])

def use_ast_literal_eval(user_input):
    """Safe alternative to eval."""
    import ast
    # SAFE: Using ast.literal_eval for simple literals
    return ast.literal_eval(user_input)

File: src/java/CommandInjection.java
package java;

import java.io.*;

/**
 * This class contains Command Injection vulnerabilities.
 * Used for testing SAST tools.
 */
public class CommandInjection {

    /**
     * Vulnerable command execution using Runtime.exec()
     */
    public void executeCommand(String userInput) throws IOException {
        // VULNERABLE: Direct command concatenation
        String command = "ping -c 1 " + userInput;
        Runtime.getRuntime().exec(command);
    }

    /**
     * Command injection with ProcessBuilder
     */
    public void executeWithProcessBuilder(String fileName) throws IOException {
        // VULNERABLE: User input in command arguments
        ProcessBuilder pb = new ProcessBuilder("cat", fileName);
        pb.start();
    }

    /**
     * Command injection with shell execution
     */
    public void executeShellCommand(String hostname) throws IOException {
        // VULNERABLE: Using shell with user input
        String[] cmd = {"/bin/sh", "-c", "nslookup " + hostname};
        Runtime.getRuntime().exec(cmd);
    }

    /**
     * Multiple command injection vectors
     */
    public void complexCommandInjection(String input) throws IOException {
        // VULNERABLE: Multiple injection points
        String cmd = "echo " + input + " | grep 'test'";
        Runtime.getRuntime().exec(new String[] {"/bin/sh", "-c", cmd});
    }

    /**
     * Safe example using whitelist validation
     */
    public void safeCommandExecution(String ipAddress) throws IOException {
        // SAFE: Validate input before use
        if (ipAddress.matches("^[0-9.]+$")) {
            String command = "ping -c 1 " + ipAddress;
            Runtime.getRuntime().exec(command);
        }
    }
}